src/qforte/circuit.cc

```diff
@@ -7,19 +7,51 @@
 #include "computer.h"
 #include "sparse_tensor.h"
 
-void Circuit::set_parameters(const std::vector<double>& params) {
-    // need a loop over only gates in state preparation circuit that
-    // have a parameter dependance (if gate_id == Rx, Ry, or Rz)
-    // TODO: make a indexing funciton using a map (Nick)
+void Circuit::set_parameters(const std::vector<std::complex<double>>& params) {
     size_t param_idx = 0;
+    size_t param_size = params.size();
     for (auto& gate : gates_) {
-        std::string gate_id = gate.gate_id();
-        if (gate_id == "Rz") {
-            size_t target_qubit = gate.target();
-            gate = make_gate(gate_id, target_qubit, target_qubit, params[param_idx]);
+        if (gate.has_parameter()) {
+            if (param_idx < param_size) {
+                size_t target_qubit = gate.target();
+                size_t control_qubit = gate.control();
+                gate = make_gate(gate.gate_id(), target_qubit, control_qubit, params[param_idx]);
+            }
             param_idx++;
         }
     }
+    if (param_idx != param_size) {
+        throw std::runtime_error("Circuit::set_parameters: number of parameters does not match "
+                                 "number of gates with parameters");
+    }
+}
+
+void Circuit::insert_gate(size_t pos, const Gate& gate) {
+    if (pos > gates_.size()) {
```

Comment:
Why don't we allow `pos = gates_.size`? Isn't that a perfectly legitimate (if inefficient) way to append a gate?

Reply:
It's legitimate. I think the code already works when `pos = gates_.size` because we only skip this step if `pos > gates_.size`.